# Домашнее задание 2 по обработке текстов

Рассмотрим задачу бинарной классификации. Пусть дано два списка имен: мужские и женские имена. Требуется разработать классификатор, который по данному имени будет определять мужское оно или женское.

Данные: 
* Женские имена: female.txt
* Мужские имена: male.txt

In [1]:
import pandas as pd

with open ('female.txt', encoding = 'utf-8') as f:
    female = f.read()
with open ('male.txt', encoding = 'utf-8') as f:
    male = f.read()

## Часть 1. Предварительная обработка данных

1. Удалите неоднозначные имена (те имена, которые являются и мужскими, и женскими дновременно), если такие есть; 
2. Создайте обучающее и тестовое множество так, чтобы в обучающем множестве классы были сбалансированы, т.е. к классу принадлежало бы одинаковое количество имен;

Предполагается, что неоднозначные мужские имена будут заканчиваться на гласные буквы, а именно 'a', 'y' или 'e', а неоднозначные женские на согласные 'x', 'g', 'm'

In [2]:
len(male.split())

2943

In [3]:
words = []
for line in open('male.txt'):
    words += [w for w in line.split() if w.endswith('a') is False and w.endswith('y') is False and w.endswith('e') is False ]
print(words)

['Aamir', 'Aaron', 'Abbot', 'Abbott', 'Abdel', 'Abdul', 'Abdulkarim', 'Abdullah', 'Abel', 'Abelard', 'Abner', 'Abraham', 'Abram', 'Adair', 'Adam', 'Adams', 'Adger', 'Adlai', 'Adnan', 'Adolf', 'Adolfo', 'Adolph', 'Adolpho', 'Adolphus', 'Adrian', 'Adrick', 'Adrien', 'Agamemnon', 'Aguinaldo', 'Agustin', 'Aharon', 'Ahmad', 'Ahmed', 'Ahmet', 'Ajai', 'Al', 'Alain', 'Alan', 'Alasdair', 'Alastair', 'Albatros', 'Albert', 'Alberto', 'Albrecht', 'Alden', 'Aldis', 'Aldo', 'Aldric', 'Aldrich', 'Aldus', 'Aldwin', 'Alec', 'Aleck', 'Alejandro', 'Aleks', 'Aleksandrs', 'Alessandro', 'Alex', 'Alexander', 'Alexei', 'Alexis', 'Alf', 'Alfonso', 'Alfonzo', 'Alford', 'Alfred', 'Alfredo', 'Algernon', 'Ali', 'Alic', 'Alister', 'Alix', 'Allah', 'Allan', 'Allen', 'Allin', 'Allyn', 'Alonso', 'Alonzo', 'Aloysius', 'Alphonso', 'Alston', 'Alton', 'Alvin', 'Alwin', 'Amadeus', 'Ambros', 'Ambrosi', 'Ambrosio', 'Ambrosius', 'Amos', 'Anatol', 'Anatollo', 'Anders', 'Andonis', 'Andreas', 'Andrej', 'Andres', 'Andrew', 'Andri

In [4]:
len(words)

2114

In [5]:
len(female.split())

5004

In [6]:
word = []
for line in open('female.txt'):
    word += [w for w in line.split() if w.endswith('x') is False and w.endswith('g') is False and w.endswith('m') is False and w.endswith('my') is False ]
print(word)

['Abagael', 'Abagail', 'Abbe', 'Abbey', 'Abbi', 'Abbie', 'Abby', 'Abigael', 'Abigail', 'Abigale', 'Abra', 'Acacia', 'Ada', 'Adah', 'Adaline', 'Adara', 'Addie', 'Addis', 'Adel', 'Adela', 'Adelaide', 'Adele', 'Adelice', 'Adelina', 'Adelind', 'Adeline', 'Adella', 'Adelle', 'Adena', 'Adey', 'Adi', 'Adiana', 'Adina', 'Adora', 'Adore', 'Adoree', 'Adorne', 'Adrea', 'Adria', 'Adriaens', 'Adrian', 'Adriana', 'Adriane', 'Adrianna', 'Adrianne', 'Adrien', 'Adriena', 'Adrienne', 'Aeriel', 'Aeriela', 'Aeriell', 'Agace', 'Agata', 'Agatha', 'Agathe', 'Aggi', 'Aggie', 'Aggy', 'Agna', 'Agnella', 'Agnes', 'Agnese', 'Agnesse', 'Agneta', 'Agnola', 'Agretha', 'Aida', 'Aidan', 'Aigneis', 'Aila', 'Aile', 'Ailee', 'Aileen', 'Ailene', 'Ailey', 'Aili', 'Ailina', 'Ailyn', 'Aime', 'Aimee', 'Aimil', 'Aina', 'Aindrea', 'Ainslee', 'Ainsley', 'Ainslie', 'Ajay', 'Alaine', 'Alameda', 'Alana', 'Alanah', 'Alane', 'Alanna', 'Alayne', 'Alberta', 'Albertina', 'Albertine', 'Albina', 'Alecia', 'Aleda', 'Aleece', 'Aleecia', 'Al

In [7]:
len(word)

4955

In [8]:
c = ['Names']
f = pd.DataFrame(word, columns=c)
m = pd.DataFrame(words, columns=c)

In [9]:
f['Sex'] = "female"
m['Sex'] = "male"

In [10]:
f_train = f.iloc[:1900, :]
f_test = f.iloc[1900:, :]
m_train = m.iloc[:1900, :]
m_test = m.iloc[1900:, :]

In [11]:
df_train = pd.concat([f_train, m_train])
df_test = pd.concat([f_test, m_test])

##  Часть 2. Базовый метод классификации

Используйте метод наивного Байеса или логистическую регрессию для классификации имен: в качестве признаков используйте символьные $n$-граммы. Сравните результаты, получаемые при разных $n=2,3,4$ по $F$-мере и аккуратности. В каких случаях метод ошибается?

Для генерации $n$-грамм используйте:

In [12]:
from nltk.util import ngrams
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import SGDClassifier, LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.decomposition import TruncatedSVD
from sklearn.feature_extraction.text import *
from sklearn.metrics import *
from sklearn.pipeline import Pipeline

#### Mетод наивного Байеса

In [13]:
Val1 = 1
Val2 = 1

clf = Pipeline([
    ('vect',CountVectorizer(analyzer='char', ngram_range=(Val1,Val2))),
    ('clf',  MultinomialNB()),
])

In [14]:
clf.fit(df_train.Names, df_train.Sex)

Pipeline(memory=None,
     steps=[('vect', CountVectorizer(analyzer='char', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=None, vocabulary=None)), ('clf', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))])

In [15]:
predicted = clf.predict(df_test.Names)

In [16]:
print("Accuracy: {0:6.2f}".format(accuracy_score(df_test.Sex, predicted)))
print("F1-measure: {0:6.2f}".format(f1_score(df_test.Sex, predicted, average='macro')))

Accuracy:   0.46
F1-measure:   0.39


Самый лучший результат достигается при n = 4
* ngram_range=(4,4) Accuracy:   0.73, F1-measure:   0.51
* ngram_range=(3, 3) Accuracy:   0.62, F1-measure:   0.48
* ngram_range=(2, 2) Accuracy:   0.56, F1-measure:   0.45
* ngram_range=(1, 1) Accuracy:   0.46, F1-measure:   0.39

#### Логистическая регрессия

In [17]:
Val1 = 2
Val2 = 2

clf1 = Pipeline([
    ('vect', CountVectorizer(analyzer='char', ngram_range=(Val1,Val2))),
    ('clf1',  LogisticRegression())
    ])

In [18]:
clf1.fit(df_train.Names, df_train.Sex)

Pipeline(memory=None,
     steps=[('vect', CountVectorizer(analyzer='char', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(2, 2), preprocessor=None, stop_words=None,
        strip...ty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False))])

In [19]:
predicted1 = clf1.predict(df_test.Names)

In [20]:
print("Accuracy: {0:6.2f}".format(accuracy_score(df_test.Sex, predicted1)))
print("F1-measure: {0:6.2f}".format(f1_score(df_test.Sex, predicted1, average='macro')))

Accuracy:   0.60
F1-measure:   0.48


Самый лучший результат достигается при n = 2
* ngram_range=(4,4) Accuracy:   0.49, F1-measure:   0.41
* ngram_range=(3, 3) Accuracy:   0.56, F1-measure:   0.45
* ngram_range=(2, 2) Accuracy:   0.60, F1-measure:   0.48
* ngram_range=(1, 1) Accuracy:   0.48, F1-measure:   0.40

Если использовать analyzer='char_wb' вместо analyzer='char', можно улучшить результат до Accuracy:   0.65, F1-measure:   0.52

##  Часть 3. Нейронная сеть


Используйте  реккурентную нейронную сеть с  LSTM для решения задачи. В ней может быть несколько слоев с LSTM, несколько слоев c Bidirectional(LSTM).  У нейронной сети один выход, определяющий класс имени. 

Представление имени для классификации в этом случае: бинарная матрица размера (количество букв в алфавите $\times$ максимальная длина имени). Обозначим его через $x$. Если первая буква имени a, то $x[1][1] = 1$, если вторая – b, то  $x[2][1] = 1$.  

Не забудьте про регуляризацию нейронной сети дропаутами. 

Сравните результаты классификации разными методами. Какой метод лучше и почему?

Сравните результаты, получаемые при разных значениях дропаута, разных числах узлов на слоях нейронной сети по $F$-мере и аккуратности. В каких случаях нейронная сеть ошибается?

Если совсем не получается запрограммировать нейронную сеть самостоятельно, обратитесь к туториалу тут: https://github.com/divamgupta/lstm-gender-predictor